### PR TITLE
server : refactor multitask handling

### PR DESCRIPTION
**Pull request by ngxson**
## Motivation

The server example has grown up in term of functionalities. I think it would be nice to start looking at how to simplify some parts and to make the code more manage-able.

The goal here is to make the http part and engine part more separated, so they can easily be debugged and probably optimized in the future.

The current architecture contains of 5 main parts:
- HTTP stack / handler
- Task queue
- Result queue
- Task dispatcher: responsible of loading tasks into slots
- Slot dispatcher: responsible of batching and calling `llama_decode`

## In this PR

Currently, multitask is added as a "super-task" that linked to other smaller tasks. However, this introduce some complexity to the task queue level. I believe that moving it to HTTP handler will simplify the task queue.

On the way doing so, I also made some small refactoring to `struct server_task` and HTTP handlers, for example:
- new `enum server_task_cmpl_type` for completion task type: normal, embedding, infill
- reuse the same http handler for both infill and completion endpoint

## Future works

I also take time to dive deeper into [httplib source code](https://github.com/yhirose/cpp-httplib) to see if there are anything else we can take advantage of. Turns out we can maybe even get rid of the notion of `deferred` tasks (in case there is no free slots to process the incoming request). I will investigate this more in another issue/PR.

In addition, one small [issue](https://github.com/ggerganov/llama.cpp/issues/9273) regarding to cancelling task also be address. However, no fix is being planned for now.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -563,6 +533,12 @@ struct server_response {
         waiting_task_ids.insert(id_task);
     }
 
+    void add_waiting_tasks(std::vector<server_task> & tasks) {
```
Comment (ggerganov) [inline examples/server/server.cpp:536]:
```suggestion
    void add_waiting_tasks(const std::vector<server_task> & tasks) {
```

File: examples/server/server.cpp [hunk 2]
```diff
@@ -571,17 +547,22 @@ struct server_response {
         waiting_task_ids.erase(id_task);
     }
 
-    // This function blocks the thread until there is a response for this id_task
-    server_task_result recv(int id_task) {
+    void remove_waiting_tasks(std::vector<server_task> & tasks) {
```
Comment (ggerganov) [inline examples/server/server.cpp]:
```suggestion
    void remove_waiting_tasks(const std::vector<server_task> & tasks) {
```

File: examples/server/server.cpp [hunk 3]
```diff
@@ -571,17 +547,22 @@ struct server_response {
...
+        for (const auto & t : tasks) {
+            remove_waiting_task_id(t.id);
+        }
+    }
+
+    // This function blocks the thread until there is a response for one of the id_tasks
+    server_task_result recv(std::unordered_set<int> & id_tasks) {
```
Comment (ggerganov) [inline examples/server/server.cpp]:
```suggestion
    server_task_result recv(const std::unordered_set<int> & id_tasks) {
```

File: examples/server/server.cpp [hunk 4]
```diff
@@ -1508,83 +1479,128 @@ struct server_context {
...
+        return tasks;
     }
 
-    void split_multiprompt_task(int id_multi, const server_task & multiprompt_task) {
-        const int prompt_count = multiprompt_task.data.at("prompt").size();
-        if (prompt_count <= 1) {
-            send_error(multiprompt_task, "error while handling multiple prompts");
-            return;
-        }
+    void cancel_tasks(std::unordered_set<int> & id_tasks) {
```
Comment (ggerganov) [inline examples/server/server.cpp]:
```suggestion
    void cancel_tasks(const std::unordered_set<int> & id_tasks) {
```

**Review by ggerganov — APPROVED** [review]
Btw, one thing that we can try to optimize is to perform the tokenization prior to enqueing the task, so that it is multi-threaded by the http threads and happens in parallel to the GPU computation. Currently, the tokenization happens inside the main loop so it is single-threaded, inbetween `llama_decode` calls.

Last time I looked into this, it got a bit tricky. But we can try to revisit it.

File: examples/server/server.cpp (hunk 3 above)
Comment (ngxson) [reply examples/server/server.cpp]: This change was added in https://github.com/ggerganov/llama.cpp/pull/9274/commits/fa2210652af7dea17e404f038c38efd745f5186a . I discovered that not having `const` will requires lambda function to be `mutable`, which cause double-free in [this CI run](https://github.com/ggerganov/llama.cpp/actions/runs/10666482802/job/29562005029) (I'm not sure why, but good to know!)

**Comment by ngxson** [comment]
@ggerganov Thanks for reviewing this PR.

FYI, currently `handle_completions` can already handle string **or** list tokens as input, so I think it will be easy to implement tokenization inside http threads. I'll have a look on this in another PR.

File: examples/server/server.cpp [hunk 5]
```diff
@@ -386,35 +386,53 @@ struct server_queue {
...
         if (task.id == -1) {
             task.id = id++;
             LOG_VERBOSE("new task id", {{"new_id", task.id}});
         }
-        queue_tasks.push_back(std::move(task));
+        if (front) {
+            queue_tasks.insert(queue_tasks.begin(), std::move(task));
+        } else {
+            queue_tasks.push_back(std::move(task));
+        }
```
[the remark this answers is not in the document]
Comment (ngxson) [reply examples/server/server.cpp]: Done in https://github.com/ggerganov/llama.cpp/pull/9274/commits/86caa35343b3554962fc9b3a35ea7ad21bd9f845

(I'm doing some benchmarks just to be sure)

**Comment by slaren** [comment]
> Btw, one thing that we can try to optimize is to perform the tokenization prior to enqueing the task, so that it is multi-threaded by the http threads and happens in parallel to the GPU computation. Currently, the tokenization happens inside the main loop so it is single-threaded, inbetween `llama_decode` calls.

If the tokenization functions are thread-safe, this needs to be clearly documented in the llama.cpp API. The general expectation is that it is thread-safe to use different objects in multiple threads, but only one thread should use the same object.

**Comment by ngxson** [comment]
(Related to current PR) I've finished some benchmarks to confirm that there is no performance lost due to this PR, so I will merge it soon to start working on the next item.

**Pull request merged**